[c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42944 [c10d][jit] initial torchbind bindings for ProcessGroup
* #47368 [WIP] use lambda to register custom class init
* #47343 [c10d] switch ProcessGroup to be managed by intrusive_ptr
* **#47075 [c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr**
* #47074 [c10d] switch Store to be managed by intrusive_ptr
* #44046 [c10d] switch ProcessGroup::Work to be managed by intrusive_ptr



Differential Revision: [D24667127](https://our.internmc.facebook.com/intern/diff/D24667127)

wanchaol committed:
[WIP][c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr"




[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr"


Differential Revision: [D24667127](https://our.internmc.facebook.com/intern/diff/D24667127)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr"




Differential Revision: [D24667127](https://our.internmc.facebook.com/intern/diff/D24667127)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -971,16 +971,17 @@ that adds a prefix to each key inserted to the store.
                   int,
-                  ::c10d::ProcessGroupNCCL::Options>(),
+                  const c10::intrusive_ptr<
+                      ::c10d::ProcessGroupNCCL::Options>&>(),
```
this is a reference of the ptr, so it means someone else need it keep it alive?

wanchaol replied:
yes, I think the caller should keep it alive. Specifically I think this option is only needed during the initialization, and we don't need to share the ownership in the `ProcessGroupNCCL` instantiation, as it only record (copy) the fields `opTimeOut` and `isHighPriorityStream`. So after construction, the `Options` is not needed anymore. 

We can also pass by pointer directly, but I didn't see if this is necessary.

pritamdamania87 replied:
For extra safety, is it fine to remove the reference to the pointer and just incur a refcount bump here?

wanchaol replied:
yep that's fine as well, I will update the PR to make this a refcount bump to ensure safety.

mrshenli reviewed (commented):
LGTM! Just have one question regarding the lifetime of options.

wanchaol committed:
Update on "[c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr"




Differential Revision: [D24667127](https://our.internmc.facebook.com/intern/diff/D24667127)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr"




Differential Revision: [D24667127](https://our.internmc.facebook.com/intern/diff/D24667127)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr"




Differential Revision: [D24667127](https://our.internmc.facebook.com/intern/diff/D24667127)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr"




Differential Revision: [D24667127](https://our.internmc.facebook.com/intern/diff/D24667127)

[ghstack-poisoned]

wanchaol committed:
Update on "[c10d] switch ProcessGroupNCCL:Options to be managed by intrusive_ptr"




Differential Revision: [D24667127](https://our.internmc.facebook.com/intern/diff/D24667127)

[ghstack-poisoned]
